File: 0090-subsets-ii/0090-subsets-ii.cpp
class Solution {
private:
    void findSubsets(int index,vector<int> & nums, vector<int> current, 
                     vector<vector<int>> &ans){
        ans.emplace_back(current);
        for(int i=index;i<nums.size();i++){
            if(i > index && nums[i] == nums[i-1]) continue;
            current.emplace_back(nums[i]);
            findSubsets(i+1,nums,current,ans);
            current.pop_back();
        }
    }
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<int> current;
        vector<vector<int>>ans;
        findSubsets(0,nums,current,ans);
        return ans;
    }
};

File: 0090-subsets-ii/0090-subsets-ii.ts
function findSubsets(nums: number[], curr: number, currSubset: number[], ans: Set<string>): void {
    if (curr >= nums.length) {
        ans.add(JSON.stringify(currSubset));
        return;
    }
    // Include the current number
    currSubset.push(nums[curr]);
    findSubsets(nums, curr + 1, currSubset, ans);
    currSubset.pop();
    // Exclude the current number
    findSubsets(nums, curr + 1, currSubset, ans);
}

function subsetsWithDup(nums: number[]): number[][] {
    let ans: Set<string> = new Set();
    let currSubset: number[] = [];
    // sorting is required to avoid duplicacy.For eg. [4 1 4] and [4 4 1] are same as elements are same.
    nums.sort((a,b)=>a-b);
    findSubsets(nums, 0, currSubset, ans);
    // Convert the stringified subsets back to arrays
    return Array.from(ans).map(subset => JSON.parse(subset));
}